Streaming writes

jfirebaugh opened the pull request:
Replace generated code that requires lots of temporary allocations with hand-written streaming writes. This improves performance significantly for sources with many features by reducing GC pressure. (Although surprisingly it seems to slow very simple sources down slightly. I'm not sure why.)

Before:

```
raw x 92,357 ops/sec ±0.88% (93 runs sampled)
simple x 97,692 ops/sec ±1.04% (94 runs sampled)
points x 114 ops/sec ±0.77% (76 runs sampled)
lots of properties x 2,861 ops/sec ±1.32% (91 runs sampled)
```

After:

```
raw x 82,764 ops/sec ±1.38% (90 runs sampled)
simple x 87,773 ops/sec ±1.05% (89 runs sampled)
points x 176 ops/sec ±1.11% (84 runs sampled)
lots of properties x 2,849 ops/sec ±1.08% (94 runs sampled)
```

Refs https://github.com/mapbox/mapbox-gl-js/issues/5208.

anandthakker commented on index.js:
```diff
@@ -39,116 +33,113 @@ function fromGeojsonVt (layers) {
+      var rings = geometry.length
+      for (var r = 0; r < rings; r++) {
+        var ring = geometry[r]
+        pbf.writeVarint(command(1, 1)) // moveto
```
See https://github.com/mapbox/vt-pbf/commit/21f0c8bf5c6affa375839a7d185d42f1983795c4#diff-168726dbe96b3ce427e7fedce31bb0bcR111, which I just merged before I noticed this PR.

mourner commented on index.js:
```diff
@@ -39,116 +33,113 @@ function fromGeojsonVt (layers) {
 
-    feature.tags = tags
-    preparedLayer.features.push(feature)
+    pbf.writeMessage(2, function () {
```
You can avoid using closures here (which have a certain overhead) if you take advantage of the third argument to `writeMessage`. This may be the root of the simple features regression.

jfirebaugh commented:
Updated and rebased.

```
raw x 91,286 ops/sec ±1.68% (88 runs sampled)
simple x 96,556 ops/sec ±1.30% (85 runs sampled)
points x 238 ops/sec ±1.88% (85 runs sampled)
lots of properties x 2,701 ops/sec ±1.41% (92 runs sampled)
```
